### ISSUE TITLE
Add stand-alone opt-in form with standard email services integrations

### ISSUE THREAD

**Issue by pippinsplugins** [AffiliateWP/AffiliateWP#2511]
@amdrew and I have discussed this extensively.

We would like to add a new stand-alone opt-in form that visitors can submit to be subscribed to the site's email service (MailChimp, ConvertKit, ActiveCampaign, etc).

The form would be displayed on a page through a shortcode and would include fields for name and email.

When submitted, the form would trigger one of two or three things:

- If a referral cookie is present, an opt-in referral will be created (see #2198)
- The email/name will be subscribed to the email service
- A customer record will be created (see #2491)

We will need to introduce a new `Opt-Ins` tab in settings that allows site admins to pick the opt-in integration (MailChimp) and to set referral amount options.

**Comment by amdrew** [comment]
It could include optional fields for `first name` and `last name`, with `email` being default. Over time, we could add support for custom fields (?), perhaps an entire forms API (?). 

In addition to the shortcode, it would (obviously) have a method to call, as well as a widget. As Gutenberg approaches, it could have a block to drag into a page/post etc.

**Comment by pippinsplugins** [comment]
@amdrew WIP but ready to begin testing PR submitted: https://github.com/AffiliateWP/AffiliateWP/pull/2528

To begin testing:

1. Checking `issue/2511`
2. Add `[opt_in]` shortcode to a page
3. Visit Affiliates > Settings > Opt-In Forms to enable an integration with MailChimp, ConvertKit, or ActiveCampaign
4. Visit the page you placed the shortcode on and submit it with and without referral cookies present.

**Comment by pippinsplugins** [comment]
Few items that have now been completed:

1. Settings for controlling which fields show in the opt-in form
2. Settings for controlling the referral amount
3. Rigorous API failure detection for integrations
4. Complete inline documentation
5. Unit tests

**Comment by amdrew** [comment]
@pippinsplugins Excellent! Will test in the morning.

**Comment by amdrew** [comment]
After selecting MC in settings and saving:

```
PHP Notice:  Undefined variable: data_center in /wp-content/plugins/AffiliateWP/includes/integrations/opt-in-platforms/class-opt-in-platform-mailchimp.php on line 24
```

**Comment by amdrew** [comment]
I'm not able to submit the form (with or without referral link). I see this after I submit the form:

<img width="841" alt="screenshot 2018-01-23 16 29 03" src="https://user-images.githubusercontent.com/52581/35256836-b0cc45da-005a-11e8-87a5-27a9b5c723f6.png">

No errors in PHP log, API key and list ID double-checked for MailChimp.

**Comment by amdrew** [comment]
Should we prefix the `[opt_in]` shortcode? I can see how it might be confusing though having it `[affiliate_opt_in]` when it's not about affiliates opting in...

**Comment by amdrew** [comment]
Was using the wrong list ID. Was using the one from the URL as opposed to the one listed in the settings. However, after changing that I now get:

<img width="876" alt="screenshot 2018-01-23 16 42 17" src="https://user-images.githubusercontent.com/52581/35257166-967cfa74-005c-11e8-805a-91213fd6b497.png">

**Comment by pippinsplugins** [comment]
Bad request error and undefined variable fixed.

**Comment by amdrew** [comment]
@pippinsplugins Can confirm the errors are gone. The form also sent my two test emails to MC 👍

We'll also need a confirmation message after submitting the opt-in form. We could probably show the confirmation like we do for other forms and show it on page reload, but it might be worth exploring ajax in the near future. The opt-in shortcode might not always be placed at the top of a page for example.

**Comment by pippinsplugins** [comment]
Confirmation notice added.

**Comment by pippinsplugins** [comment]
ConvertKit integration now working.

**Comment by amdrew** [comment]
@pippinsplugins ConvertKit works perfectly 👍

**Comment by pippinsplugins** [comment]
@amdrew what else would you like to see here for MVP? Do you think we can get by with not having options to select the required / shown fields or are those required for an MVP?

**Comment by amdrew** [comment]
@pippinsplugins In my 2 tests, both subscribers were added to ActiveCampaign, found them under `All contacts`. I don't see them on my actual list though, I wonder if AC just needs time to update.

**Comment by amdrew** [comment]
@pippinsplugins I think we can skip the First name and Last name field options for now. 

At the very minimum we need an input field for specifying an opt-in referral rate. I noticed it was completed as per above (https://github.com/AffiliateWP/AffiliateWP/issues/2511#issuecomment-359301031) but I can't see it anywhere.

**Comment by pippinsplugins** [comment]
@amdrew whoops, missed that! New setting added for defining the referral amount.

**Comment by pippinsplugins** [comment]
@gspice @michaelbeil could both of you run through tests on here too please?

To test:

1. Checkout branch `issue/2511`
2. Go to Affiliates > Settings > Opt-In Form
3. Select the platform to connect (MailChimp, ConvertKit, and ActiveCampaign available)
4. Save settings
5. Enter your API key, API URL (if applicable), and list ID.
6. Add `[opt_in`] shortcode to a page
7. Visit the page on a fresh browser _without_ a referral URL, and submit the form. The email entered should be subscribed to the opt-in platform but no referral should be created.
8. Repeat step 7 but this time visit the page with a referral variable present in the URL.
9. Submit the form and you should now get opt-in success _and_ a $0.00 referral created
10. Enter an opt-in referral amount in `Opt-In Form` settings and repeat steps 8 and 9 to confirm the amount matches.

**Comment by amdrew** [comment]
@pippinsplugins 

In the `opt-in.php` template file there are these hooks:

affwp_affiliate_opt_in_form_top
affwp_affiliate_opt_in_form_bottom

Can we take away the `affiliate` so they are consistent with the other hook names in the same file?

affwp_opt_in_fields_before
affwp_opt_in_fields_after

**Comment by amdrew** [comment]
@pippinsplugins Could we change the referral amount field to a `number` field and make it have a `step` of something like `0.1`? Would make it more consistent with the other referral rate field on the `General` tab.

**Comment by amdrew** [comment]
@pippinsplugins Thoughts on a filter here? https://github.com/AffiliateWP/AffiliateWP/blob/issue/2511/includes/integrations/class-opt-in.php#L132

I could see someone wanting to change the default status (but probably a bad idea) so they don't need to manually approve every single opt-in. Also being able to change the default `type` or `description` (they could make up their own) would be neat.

**Comment by pippinsplugins** [comment]
Note to self: need to add a setting to control the opt-in success message.

**Comment by gspice** [comment]
I love this new feature!

Tested with MailChimp list:
* Successful in submitting opt-in form, adding to MC list w/out using an affiliate link.
* Referral was created when submitting opt-in form **with** an affiliate link in both zero and the correct dollar amounts when populated.  

Notes:
* Visit was not changed to **Converted** when referral was created, possibly from next item...
* Referral was created as **Pending** and when I clicked **Accept** I received this warning:

![warning](https://user-images.githubusercontent.com/3018245/36131095-07a8bef4-1025-11e8-8a8b-fbca00325996.png)

**Comment by amdrew** [comment]
We've decided to  add 2 new settings:

1. Double opt-in
2. Default referral status (pending (default) and unpaid)

**Comment by amdrew** [comment]
@gspice ActiveCampaign has a 2 week trial. Added ConvertKit to our Trello card.

**Comment by pippinsplugins** [comment]
The error message @gspice saw was due to the email settings that got accidentally merged into `release/2.2` (which this branch is based off of). We'll need to get #948 merged in before addressing it.

**Comment by pippinsplugins** [comment]
Status option added.

**Comment by amdrew** [comment]
Status option works great.

@pippinsplugins I wonder if we should hide the form once it has been submitted:

<img width="623" alt="screen shot 2018-02-20 at 9 38 14 am" src="https://user-images.githubusercontent.com/52581/36396327-60c8d5a2-1622-11e8-9f38-1a7f2d2630b7.png">

1. I'm pretty sure this is how other opt-in forms behave within other plugins
2. It would help (just a little) with someone submitting another email address and a new referral being created.

**Comment by gspice** [comment]
I've re-tested with all 3 integration.  It successfully creates referrals as expected with the correct status that was selected in settings.  I'm still getting that warning if the status is **Pending**, and I click to **Accept**, but you probably already know that.

**Comment by pippinsplugins** [comment]
Form now hidden on success.

**Comment by amdrew** [comment]
@michaelbeil I had that issue here: https://github.com/AffiliateWP/AffiliateWP/issues/2511#issuecomment-359665102

Is your issue/2511 branch up to date with the latest commits? That error should have been fixed and the form should disappear after submission, I wonder if that's why. Which platform are you testing?

**Comment by amdrew** [comment]
@pippinsplugins If you forget to put a last name, the fields are all wiped and you need to re-enter them. We should probably keep the existing values like we do on the registration form.

**Comment by gspice** [comment]
@m I think I received that `Not Found` error my first time testing, turns out I had the wrong number entered for my MailChimp list id.

**Comment by amdrew** [comment]
That's right, I did the same.

@michaelbeil See https://github.com/AffiliateWP/AffiliateWP/issues/2511#issuecomment-359667241

Double-check that your list ID is the one from the settings of the list, like this one on the right side:

<img width="1054" alt="screen shot 2018-02-27 at 7 31 42 am" src="https://user-images.githubusercontent.com/52581/36688099-51fea718-1b90-11e8-8a78-850948eb3cbe.png">

**Comment by michaelbeil** [comment]
That works now!

**Comment by amdrew** [comment]
@michaelbeil Did you get a chance to test with the other email services?

**Comment by michaelbeil** [comment]
Referrals were generated successfully for both ActiveCampaign and ConvertKit. However, on the visits table, these visits do not show as converted for ConvertKit:

![screen shot on 2018-03-02 at 18_18_33](https://user-images.githubusercontent.com/4752949/36912228-6c6565fc-1e46-11e8-9818-2ade10ac40ae.png)
![screen shot on 2018-03-02 at 18_19_17](https://user-images.githubusercontent.com/4752949/36912233-7208670c-1e46-11e8-858c-da160e0d813b.png)

**Comment by amdrew** [comment]
@michaelbeil And visits _are_ showing as converted for ActiveCampaign and MailChimp?

I'm not seeing visits being converted for any of the 3 integrations.

**Comment by pippinsplugins** [comment]
@amdrew do you have your default referral status set to unpaid?

**Comment by amdrew** [comment]
@pippinsplugins Yes

**Comment by michaelbeil** [comment]
Visits are _not_ showing as converted for all 3 integrations.

**Comment by pippinsplugins** [comment]
@michaelbeil @amdrew try again now please.

**Comment by amdrew** [comment]
@pippinsplugins Not working for me:

```
PHP Notice:  Undefined variable: referral in /wp-content/plugins/AffiliateWP/includes/integrations/class-opt-in.php on line 145

PHP Notice:  Trying to get property 'visit_id' of non-object in /wp-content/plugins/AffiliateWP/includes/integrations/class-opt-in.php on line 145
```

**Comment by pippinsplugins** [comment]
Oops, should be fixed now.

**Comment by amdrew** [comment]
@pippinsplugins 👍Visits now show as converted for all 3 integrations

**Comment by michaelbeil** [comment]
Everything looks good now with visits showing as converted!

**Comment by amdrew** [comment]
@pippinsplugins It would be great to get this merged soon into #2491 so we can test the 3rd item in the original issue description above:

> A customer record will be created 

Or maybe merged into 2.2 and then we can update 2491 from that, not sure what the best option is.

This would also allow us to test Lifetime Commissions, since a customer and an affiliate should be linked (https://github.com/AffiliateWP/affiliate-wp-lifetime-commissions/issues/61) via the new customers database table (https://github.com/AffiliateWP/affiliate-wp-lifetime-commissions/issues/75).

**Comment by pippinsplugins** [comment]
@amdrew My plan is to try and get the majority of the remaining work done on this this week.

**Comment by pippinsplugins** [comment]
@amdrew once I finish up a few more items on #2491, I'll get that merged into `release/2.2` then updated in this branch so we can finish out this.

**Comment by amdrew** [comment]
👍 Sounds good

**Comment by pippinsplugins** [comment]
#2491 is merged into release/2.2 and t his branch has been brought up to date.

**Comment by pippinsplugins** [comment]
A double opt-in setting has been added for MailChimp. Neither ConvertKit nor ActiveCampaign support setting double opt-in through the API; they only support it as a list setting inside their dashboards.

**Comment by pippinsplugins** [comment]
@amdrew all items addressed! Give it another run?

**Comment by amdrew** [comment]
@pippinsplugins Everything works great so far except it looks like referrals aren't being generated for me anymore (maybe because of the merge):

```
WordPress database error Unknown column 'customer_id' in 'field list' for query INSERT INTO `wp_77_affiliate_wp_referrals` (`affiliate_id`, `customer_id`, `date`, `currency`, `type`, `status`, `amount`, `description`, `visit_id`, `reference`) VALUES (48, 0, '2018-03-28 21:30:50', 'USD', 'opt-in', 'pending', '0.07', 'Andrew Munro', 88, 'andrew+test26@sumobi.com') made by require('wp-blog-header.php'), require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('init'), WP_Hook->do_action, WP_Hook->apply_filters, affwp_do_actions, do_action('affwp_opt_in'), WP_Hook->do_action, WP_Hook->apply_filters, AFFWP\Integrations\Opt_In->process_opt_in, Affiliate_WP_Referrals_DB->add, Affiliate_WP_DB->insert
```

**Comment by DrewAPicture** [comment]
Definitely didn't intend to close it. Whoops!

**Comment by pippinsplugins** [comment]
@amdrew can you re-run the upgrade routine or manually update the DB? That means your upgrade routine probably ran before I added the `customer_id` column to referrals.

**Comment by amdrew** [comment]
@pippinsplugins All working well now.

I propose we drop the `affwp-notice` div around the success message. When submitting my email address I would expect to see just a thanks message, not a "notice" of any sort.

Depending on the size of the message the site admin adds, the formatting they use from the editor, and the existing styling of the theme for things like heading color etc, it could cause styling issues. If the site admin also wanted to use this to offer an upsell or similar, they would need to now add more CSS to remove the notice styling.

**Comment by pippinsplugins** [comment]
What class name would you like instead?

**Comment by amdrew** [comment]
@pippinsplugins How about `affwp-opt-in-success-message`?

**Comment by amdrew** [comment]
After that, I really think we're good to go on this issue 🎉

**Comment by amdrew** [comment]
@pippinsplugins Actually, I standby my earlier comment - let's drop the div altogether, we don't need it. We should instead make the editor output tags. 

I.e. someone might do the following, since we have provided them with a visual editor:

<img width="1231" alt="screen shot 2018-04-03 at 3 21 19 pm" src="https://user-images.githubusercontent.com/52581/38227569-e4c347a6-3752-11e8-9130-f9792c5f0ffd.png">

It would be expected that the above outputs 3 paragraphs and an image but instead we get no tags:

<img width="830" alt="screen shot 2018-04-03 at 3 21 11 pm" src="https://user-images.githubusercontent.com/52581/38227579-faf6811e-3752-11e8-8442-e4ddb3624265.png">

**Comment by pippinsplugins** [comment]
@amdrew Updated!

**Comment by pippinsplugins** [comment]
@amdrew The exact error doesn't seem possible to detect (at least not that I've discovered yet) so crafting "pretty" error messages for things like `Email address already subscribed` may not be doable. I have, however, updated MailChimp to detect when a verbose error is present and that is now shown, so now we have:

![image](https://user-images.githubusercontent.com/1034109/38531709-db169fa8-3c36-11e8-8c64-1abbc7803779.png)

**Comment by amdrew** [comment]
@pippinsplugins MailChimp for WordPress has a configurable message so it must be possible.

<img width="821" alt="screen shot 2018-04-11 at 11 03 17 am" src="https://user-images.githubusercontent.com/52581/38588188-af1a3dd6-3d78-11e8-938b-fc8c779f3a07.png">

The `Use PUT to..` portion of that message isn't something we would want to show for a user subscribing.

Unless I'm mistaken, it looks like you can check for an existing user through ActiveCampaign's API:
https://www.activecampaign.com/blog/adding-or-updating-an-existing-subscriber-through-the-api/

Probably something similar for ConvertKit though I haven't looked.

**Comment by pippinsplugins** [comment]
That's handled, I'm certain, through an additional API call to first detect if the email is subscribed before subscribing it. I was hoping to avoid that but I think that might be the best way.

**Comment by pippinsplugins** [comment]
@amdrew Please try again now. I've added an `already subscribed` check that runs prior to attempting to subscribe an email.

**Comment by pippinsplugins** [comment]
Note: ConvertKit doesn't have a method for this that supports checking a specific email address. They require the subscriber ID, which we won't have at this time.

**Comment by amdrew** [comment]
@pippinsplugins Perfect!

I think we're good to go?

**Comment by pippinsplugins** [comment]
@amdrew going to get #2198 finished and merged first. Just waiting on final testing.

**Comment by pippinsplugins** [comment]
Once tests pass again, I'll merge this in. Then it's just a matter of writing docs and releasing beta!

**Comment by pippinsplugins** [comment]
Merged!

Will close once this doc is finished: https://secure.helpscout.net/docs/5463d15be4b0f639418397d0/article/5ad8fbad0428630750929c93

**Comment by pippinsplugins** [comment]
@gspice could you please write up the documentation for this this week?

**Comment by pippinsplugins** [comment]
Documentation has been completed and dis ready to publish upon release.

**Comment by michaelbeil** [comment]
Presale questions regarding this feature: https://secure.helpscout.net/conversation/574822239/81637?folderId=634609.

I will get in touch with her when this feature is released.
